staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -116,20 +96,52 @@ func WithAudit(handler http.Handler, requestContextMapper request.RequestContext
 			if ev.ResponseStatus == nil && longRunningSink != nil {
 				ev.ResponseStatus = fakedSuccessStatus
 				ev.Stage = auditinternal.StageResponseStarted
-				processEvent(longRunningSink, ev)
+				processAuditEvent(longRunningSink, ev)
 			}
 
 			ev.Stage = auditinternal.StageResponseComplete
 			if ev.ResponseStatus == nil {
 				ev.ResponseStatus = fakedSuccessStatus
 			}
-			processEvent(sink, ev)
+			processAuditEvent(sink, ev)
 		}()
 		handler.ServeHTTP(respWriter, req)
 	})
 }
 
-func processEvent(sink audit.Sink, ev *auditinternal.Event) {
+func createAuditEvent(ctx request.Context, policy policy.Checker, requestContextMapper request.RequestContextMapper, req *http.Request, w http.ResponseWriter) (*auditinternal.Event, error) {
+	attribs, err := GetAuthorizerAttributes(ctx)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to GetAuthorizerAttributes: %v", err))
+		responsewriters.InternalError(w, req, errors.New("failed to parse request"))
+		return nil, err
+	}
+
+	level := policy.Level(attribs)
+	audit.ObservePolicyLevel(level)
+	if level == auditinternal.LevelNone {
+		// Don't audit.
+		return nil, nil
+	}
+
+	ev, err := audit.NewEventFromRequest(req, level, attribs)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to complete audit event from request: %v", err))
+		responsewriters.InternalError(w, req, errors.New("failed to update context"))
+		return nil, err
+	}
+
+	ctx = request.WithAuditEvent(ctx, ev)
```

Comment:
this is unexpected in a `create*` function as a side-effect.

Reply:
either move that our or call the func createAuditEventAndAttachToCtx

Reply:
I moved context extraction also to that method and went with `createAuditEventAndAttachToContext`.